### PR TITLE
[MB-2603] Create pricer for destination SIT delivery

### PR DESCRIPTION
**Pull request by reggieriser**
## Description

This PR creates a pricer for the destination SIT delivery service item (`DDDSIT`).

## Reviewer Notes

I made a couple of related refactorings along the way:
- I changed the signature of the domestic linehaul pricer to use `unit.Pound` and `unit.Miles` to match the other pricers (this pricer leverages linehaul in some cases).
- I found that we had some cases where the `RequestedPickupDate` parameter was being configured as a timestamp rather than a date in our test data setup.
- I adjusted the EDI generation/tests to take into account the new unique distance keys used by `DDDSIT` (and did `DOPSIT` while I was at it).
- I added a couple of generated mocks that were missing from another pricer's PR.

## Setup

Given the sometimes incomplete state of some of our E2E data, let's start with a new move.  Spin up both the server and the client:

`make server_run`
`make client_run`

Then go to `http://milmovelocal:3000` and create a new HHG move with a shipment.  The origin duty station's GBLOC must be in the same GBLOC as the office user you look at it with (see this [Wiki page](https://github.com/transcom/mymove/wiki/How-to-view-a-move-or-payment-request-in-the-office-app-as-a-TOO-or-TIO) for more details).  I chose `Yuma AFB`.  Take note of the move locator on the last page.

Go to `http://officelocal:3000` and find the move you just created via the move locator.  Go to the orders and set any empty fields like TAC/SAC.  Go ahead and approve the move and the default service items for shipment management and counseling fees.

Now lets use the `prime-api-demo` script to handle some of the usual things like updating dates/weights that the prime would be doing.  Run these commands and complete it up until the point of creating a payment request, then stop the script:

```
make bin/prime-api-client
prime-api-demo <insert move ID> local
```

Now we need to add a `DDFSIT` (destination first day SIT) to the move.  This will also cause two other SIT codes to be added automatically, included the `DDDSIT` that is priced by this PR.  Use this JSON template to make a file named `create_ddfsit_service_item.json`, filling in the move ID and shipment ID (which you can find in your local database from the locator):

```json
{
  "body": {
    "moveTaskOrderID": "<insert move ID>",
    "mtoShipmentID": "<insert shipment ID>",
    "modelType": "MTOServiceItemDestSIT",
    "reServiceCode": "DDFSIT",
    "timeMilitary1": "1705Z",
    "firstAvailableDeliveryDate1": "2021-01-31",
    "timeMilitary2": "0719Z",
    "firstAvailableDeliveryDate2": "2021-02-03",
    "sitEntryDate": "2021-01-12",
    "sitDepartureDate": "2021-01-22"
  }
}
```

Create the `DDDSIT` service item with this command (which also sets up the next command to update a SIT final destination address):

```
prime-api-client --insecure create-mto-service-item --filename create_ddfsit_service_item.json | jq "map(select(.reServiceCode == \"DDDSIT\"))[0] | {mtoServiceItemId: .id, ifMatch: .eTag, body: {id: .id, modelType: \"UpdateMTOServiceItemSIT\", sitDestinationFinalAddress: {streetAddress1: \"20245 4th St\", city: \"Albuquerque\", state: \"NM\", postalCode: \"87117\", country: \"United States\"}}}" > update_dddsit_service_item.json
```

You should now have a `update_dddsit_service_item.json` file and can issue this command to set the SIT final destination address:

```
prime-api-client --insecure update-mto-service-item --filename update_dddsit_service_item.json | jq
```

Now we should be able to issue a payment request for the `DDDSIT` service item.  Create a `create_payment_request.json` file with these contents (looking up the IDs noted below in your database):

```json
{
  "body": {
    "isFinal": false,
    "moveTaskOrderID": "<insert MTO ID>",
    "serviceItems": [
      {
        "id": "<insert DDDSIT MTO service item ID>"
      }
    ]
  }
}
```

Now send this payload to create the payment request:

```
prime-api-client create-payment-request --insecure --filename create_payment_request.json | jq
```

If all goes well, the payload returned should show all the parameters associated with the `DDDSIT` along with the price.  See the [SIT pricing notes](https://docs.google.com/document/d/1jRl91aslMnOMt2zS1tt_XT0lD1I-xvsiKOfprAaFrAs/edit) or [this Wiki page](https://github.com/transcom/mymove/wiki/Acceptance-Testing-Payment-Requests#domestic-destination-sit-delivery-dddsit) for details of how each of those are supposed to be priced.  Compare your calculations to what was returned in the payload.

As you see in the SIT pricing notes, there are several different ways SIT destination may be priced:

1.  Zip3 to same zip3
2. Zip3 to different zip3 and > 50 miles
3. Zip3 to different zip3 and <= 50 miles

You may want to test each of these scenarios with representative zips/distances.

### Acceptance Instructions

Acceptance should look very similar to the above, but using the `prime-api-demo` and `prime-api-client` to talk to staging instead.  You may also need to use the `fetch-mto-updates` command of the `prime-api-client` to get the required IDs and/or eTags since we can't see the staging database directly.

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2603) for this change
* [SIT Pricing Notes](https://docs.google.com/document/d/1jRl91aslMnOMt2zS1tt_XT0lD1I-xvsiKOfprAaFrAs/edit)

### CODE REVIEW

File: pkg/services/ghcrateengine/domestic_destination_sit_delivery_pricer.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+func (p domesticDestinationSITDeliveryPricer) Price(contractCode string, requestedPickupDate time.Time, isPeakPeriod bool,
+	weight unit.Pound, serviceArea string, sitSchedule int, zipDest string, zipSITDest string, distance unit.Miles) (unit.Cents, error) {
+
+	if weight < minDomesticWeight {
+		return 0, fmt.Errorf("weight of %d less than the minimum of %d", weight, minDomesticWeight)
+	}
+
+	// Three different pricing scenarios below.
+
+	// 1) Less than or equal to 50 miles (if same or different zip3s)
```
Comment (jacquelineIO) [inline pkg/services/ghcrateengine/domestic_destination_sit_delivery_pricer.go:36]: Left thread in Slack to discuss. I don't think this is the correction logic.

**Comment by pearl-truss** [comment]
Does this work if you try to generate an EDI with the resulting payment request?

File: pkg/services/ghcrateengine/domestic_destination_sit_delivery_pricer.go (hunk 1 above)
Comment (reggieriser) [reply pkg/services/ghcrateengine/domestic_destination_sit_delivery_pricer.go]: Adjusted pricing pathways logic.

**Comment by reggieriser** [comment]
> Does this work if you try to generate an EDI with the resulting payment request?

No, it did fail due to the EDI generation code looking for distance keys that didn't apply to `DDDSIT`.  I've made changes to the EDI generation and tests to use the appropriate keys now (and did `DOPSIT` too).

**Comment by pearl-truss** [comment]
> > Does this work if you try to generate an EDI with the resulting payment request?
> 
> No, it did fail due to the EDI generation code looking for distance keys that didn't apply to `DDDSIT`. I've made changes to the EDI generation and tests to use the appropriate keys now (and did `DOPSIT` too).

I was able to see that it got priced but I got an error generating the EDI. It might matter -the payment service item for DistanceZipSITDest was 539.

**Review by jacquelineIO — APPROVED** [review]
lgtm. I have reviewed the 3 pricer paths locally in dev. thanks for all the testing and work here. I know there might still be a merge left. Up to you if you still need an app-eng review or not.

**Review by lynzt — APPROVED** [review]
I only looked at the migration, but the migration LGTM

**Comment by pearl-truss** [comment]
I just re-ran it and it worked for me!

**Pull request merged**